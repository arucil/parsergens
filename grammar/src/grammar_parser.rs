
pub mod lex;
pub mod ast;

use lalrpop_util::{lalrpop_mod, ParseError};

lalrpop_mod!(pub grammar, "/grammar_parser/grammar.rs");

pub fn parse(
  input: &str
) -> Result<ast::Grammar, ParseError<usize, lex::Token, lex::LexError>> {
  let lexer = lex::Lexer::new(input);
  grammar::DocumentParser::new().parse(input, lexer)
}
