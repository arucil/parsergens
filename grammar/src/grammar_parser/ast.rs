use std::ops::Range;

pub type Grammar = Vec<Spanned<Decl>>;

#[derive(Debug)]
pub enum Decl {
  Token(TokenDecl),
  Start(StartDecl),
  Rule(RuleDecl),
  Skip(SkipDecl),
}

#[derive(Debug)]
pub struct TokenDecl {
  pub name: Spanned<String>,
  pub pattern: TokenPattern,
}

#[derive(Debug)]
pub enum TokenPattern {
  Regex(Spanned<(String, Regex)>),
  String(Spanned<(String, Regex)>),
}

#[derive(Debug)]
pub enum Regex {
  Any,
  Char(char),
  CharSet(Vec<Range<char>>),
  CharClass(CharClassKind),
  Or(Box<Regex>, Box<Regex>),
  Optional(Box<Regex>),
  Many(Box<Regex>),
  Many1(Box<Regex>),
  Seq(Vec<Regex>),
}

#[derive(Debug)]
pub enum CharClassKind {
  Digit,
  Word,
}

#[derive(Debug)]
pub struct SkipDecl {
  pub pattern: TokenPattern,
}

#[derive(Debug)]
pub struct StartDecl {
  pub name: Spanned<String>,
}

#[derive(Debug)]
pub struct RuleDecl {
  pub name: Spanned<String>,
  pub alts: Vec<Spanned<RuleAlt>>,
}

#[derive(Debug)]
pub enum RuleAlt {
  Epsilon,
  Terms(Vec<Spanned<String>>),
}

#[derive(Debug)]
pub struct Spanned<T>(pub Range<usize>, pub T);