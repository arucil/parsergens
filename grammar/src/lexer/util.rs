
pub fn find_char_interval(char: u32, char_intervals: &[u32]) -> u32 {
  let mut lo = 0;
  let mut hi = char_intervals.len();

  while lo < hi {
    let mid = (lo + hi) / 2;
    if char_intervals[mid] > char {
      hi = mid;
    } else {
      lo = mid + 1;
    }
  }

  lo as u32 - 1
}
