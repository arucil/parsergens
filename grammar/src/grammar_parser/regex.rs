use std::str::CharIndices;
use std::iter::Peekable;
use either::Either::{self, *};
use super::ast::{Regex, CharClass, CharSetItem};

#[derive(Debug)]
pub struct RegexError {
  pub kind: RegexErrorKind,
  pub span: (usize, usize),
}

#[derive(Debug)]
pub enum RegexErrorKind {
  SyntaxError,
  InvalidCodePoint,
  Empty,
}

type Result<T=Regex> = std::result::Result<T, RegexError>;

pub fn parse_string_literal(
  input: &str,
  start: usize,
) -> Result {
  RegexParser::new(input)
    .parse_string_literal()
    .map_err(|err| {
      RegexError {
        kind: err.kind,
        span: (err.span.0 + start, err.span.1 + start),
      }
    })
}

pub fn parse_regex(
  input: &str,
  start: usize,
) -> Result {
  RegexParser::new(input)
    .parse()
    .map_err(|err| {
      RegexError {
        kind: err.kind,
        span: (err.span.0 + start, err.span.1 + start),
      }
    })
}

struct RegexParser<'a> {
  input: &'a str,
  chars: Peekable<CharIndices<'a>>,
}

impl<'a> RegexParser<'a> {
  fn new(input: &'a str) -> Self {
    Self {
      input,
      chars: input.char_indices().peekable()
    }
  }

  fn parse(&mut self) -> Result {
    let regex = self.parse_regex()?;

    if let Some((i, _)) = self.chars.next() {
      syntax_error(i, self.input.len())?
    } else {
      Ok(regex)
    }
  }

  fn parse_string_literal(&mut self) -> Result {
    let mut items = vec![];

    while let Some((i0, c)) = self.chars.next() {
      if c == '\\' {
        match self.parse_escape(i0, false)? {
          Left(c) => items.push(Regex::Char(c)),
          _ => unreachable!(),
        }
      } else {
        items.push(Regex::Char(c));
      }
    }

    if items.is_empty() {
      Ok(Regex::Empty)
    } else if items.len() == 1 {
      Ok(items.pop().unwrap())
    } else {
      Ok(Regex::Concat(items))
    }
  }

  fn parse_regex(&mut self) -> Result {
    self.parse_alt()
  }

  fn parse_alt(&mut self) -> Result {
    let mut items = vec![self.parse_concat()?];

    while let Some((_, '|')) = self.chars.peek() {
      self.advance();
      items.push(self.parse_concat()?);
    }

    if items.len() == 1 {
      Ok(items.pop().unwrap())
    } else {
      Ok(Regex::Alt(items))
    }
  }

  fn parse_concat(&mut self) -> Result {
    let mut items = vec![self.parse_repetition()?];

    loop {
      if let Some((_, '|' | ')')) | None = self.chars.peek() {
        break;
      }

      items.push(self.parse_repetition()?);
    }

    if items.len() == 1 {
      Ok(items.pop().unwrap())
    } else {
      Ok(Regex::Concat(items))
    }
  }

  fn parse_repetition(&mut self) -> Result {
    let mut item = self.parse_factor()?;

    while let Some(&(_, c@('+' | '*' | '?'))) = self.chars.peek() {
      self.advance();
      item = match c {
        '+' => Regex::Many1(Box::new(item)),
        '*' => Regex::Many(Box::new(item)),
        '?' => Regex::Optional(Box::new(item)),
        _ => unreachable!(),
      };
    }

    Ok(item)
  }

  fn parse_factor(&mut self) -> Result {
    match self.chars.peek() {
      Some((_, '.')) => {
        self.advance();
        Ok(Regex::Any)
      }
      Some((_, '(')) => {
        self.advance();
        let regex = self.parse_regex()?;
        self.match_char(')')?;
        Ok(regex)
      }
      Some(&(i, '\\')) => {
        self.advance();
        match self.parse_escape(i, true)? {
          Left(c) => Ok(Regex::Char(c)),
          Right(class) => Ok(Regex::CharClass(class)),
        }
      }
      Some(&(i, '[')) => {
        self.advance();
        self.parse_char_set(i)
      }
      Some(&(i, c@('+' | '*' | '?'))) => {
        syntax_error(i, i + c.len_utf8())?
      }
      Some((_, '|' | ')')) | None => {
        Ok(Regex::Empty)
      }
      Some(&(_, c)) => {
        self.advance();
        Ok(Regex::Char(c))
      }
    }
  }

  fn parse_char_set(&mut self, _start: usize) -> Result {
    let mut pending_range = false;
    let mut items = vec![];

    loop {
      let item = match self.chars.next() {
        Some((_, ']')) => {
          if pending_range {
            items.push(CharSetItem::Char('-'));
          }
          return Ok(Regex::CharSet(items));
        }
        Some((i, '\\')) => {
          match self.parse_escape(i, true)? {
            Left(c) => CharSetItem::Char(c),
            Right(class) => CharSetItem::CharClass(class),
          }
        }
        Some((_, '-')) => {
          if pending_range {
            CharSetItem::Char('-')
          } else if let Some(CharSetItem::Char(_)) = items.last() {
            pending_range = true;
            continue;
          } else {
            CharSetItem::Char('-')
          }
        }
        Some((_, c)) => {
          CharSetItem::Char(c)
        }
        None => {
          self.eof_error()?
        }
      };

      if pending_range {
        match &item {
          CharSetItem::Char(c) => {
            let c0 = match items.pop().unwrap() {
              CharSetItem::Char(c) => c,
              _ => unreachable!(),
            };
            items.push(CharSetItem::Range(c0, *c));
          }
          _ => {
            items.push(CharSetItem::Char('-'));
            items.push(item);
          }
        }
        pending_range = false;
      } else {
        items.push(item);
      }
    }
  }

  fn parse_escape(
    &mut self,
    start: usize,
    enable_class: bool
  ) -> Result<Either<char, CharClass>> {
    match self.chars.next() {
      Some((_, 'n')) => {
        Ok(Left('\n'))
      }
      Some((_, 'r')) => {
        Ok(Left('\r'))
      }
      Some((_, 't')) => {
        Ok(Left('\t'))
      }
      Some((_, '\\')) => {
        Ok(Left('\\'))
      }
      Some((_, 'd')) if enable_class => {
        Ok(Right(CharClass::Digit))
      }
      Some((_, 'w')) if enable_class => {
        Ok(Right(CharClass::Word))
      }
      Some((i, 'U')) => {
        if let Some((_, '{')) = self.chars.next() {
          ()
        } else {
          syntax_error(start, i + 1)?
        }

        let cp_start = i + 2;
        let cp_end;
        loop {
          match self.chars.next() {
            Some((i, '}')) => {
              cp_end = i;
              break;
            }
            None => {
              syntax_error(start, self.input.len())?
            }
            _ => {}
          }
        }

        if cp_start == cp_end {
          syntax_error(start, cp_end)?
        } else {
          let ch = u32::from_str_radix(&self.input[cp_start..cp_end], 16)
            .map_err(|_| ())
            .and_then(|cp| std::char::from_u32(cp).ok_or(()))
            .map_err(|_| RegexError {
              kind: RegexErrorKind::InvalidCodePoint,
              span: (cp_start, cp_end),
            })?;

          Ok(Left(ch))
        }
      }
      Some((_, c)) => {
        Ok(Left(c))
      }
      None => {
        self.eof_error()?
      }
    }
  }

  fn match_char(&mut self, c: char) -> Result<char> {
    self.match_char_if(|x| x == c)
  }

  fn match_char_if(&mut self, f: impl Fn(char) -> bool) -> Result<char> {
    if let Some((i, c)) = self.chars.next() {
      if f(c) {
        Ok(c)
      } else {
        syntax_error(i, i + c.len_utf8())?
      }
    } else {
      self.eof_error()?
    }
  }

  fn advance(&mut self) {
    self.chars.next();
  }

  fn eof_error(&self) -> Result<!> {
    Err(RegexError {
      kind: RegexErrorKind::SyntaxError,
      span: (self.input.len(), self.input.len())
    })
  }
}

fn syntax_error(start: usize, end: usize) -> Result<!> {
  Err(RegexError {
    kind: RegexErrorKind::SyntaxError,
    span: (start, end),
  })
}


#[cfg(test)]
mod tests {
  use super::*;
  use insta::assert_debug_snapshot;

  mod string_literal {
    use super::*;

    #[test]
    fn trivial() {
      assert_debug_snapshot!(parse_string_literal("a.-+&Ro1", 100));
    }

    #[test]
    fn empty() {
      assert_debug_snapshot!(parse_string_literal("", 100));
    }

    #[test]
    fn single_char() {
      assert_debug_snapshot!(parse_string_literal("+", 100));
    }

    #[test]
    fn escape() {
      assert_debug_snapshot!(parse_string_literal(r"ab\U{41}\r\nn\tt\d", 100));
    }
  }

  mod regex {
    use super::*;

    #[test]
    fn empty() {
      assert_debug_snapshot!(parse_regex(r"", 100));
    }

    #[test]
    fn trivial() {
      assert_debug_snapshot!(parse_regex("aRo1", 100));
    }

    #[test]
    fn escape() {
      assert_debug_snapshot!(parse_regex(r"ab\U{41}\r\nn\tt\.\+\d", 100));
    }

    #[test]
    fn alt_empty() {
      assert_debug_snapshot!(parse_regex(r"||ab", 100));
    }

    #[test]
    fn concat_precedes_alt() {
      assert_debug_snapshot!(parse_regex(r"A1\.|\d|ab", 100));
    }

    #[test]
    fn repetition_precedence() {
      assert_debug_snapshot!(parse_regex(r"ab+\||.*|ab?|", 100));
    }

    #[test]
    fn multi_repetition() {
      assert_debug_snapshot!(parse_regex(r"ab+?|\d+*", 100));
    }

    #[test]
    fn parenthesis() {
      assert_debug_snapshot!(parse_regex(r"(ab|\w\d+)+|\.", 100));
    }

    #[test]
    fn char_set_trivial() {
      assert_debug_snapshot!(parse_regex(r"[1234567890]+", 100));
    }

    #[test]
    fn char_set_range() {
      assert_debug_snapshot!(parse_regex(r"[0-9]+", 100));
    }

    #[test]
    fn char_set_multi_range() {
      assert_debug_snapshot!(parse_regex(r"[0-9a-zA-Z]+", 100));
    }

    #[test]
    fn char_set_mixed() {
      assert_debug_snapshot!(parse_regex(r"[0-9abcA-Z.]+", 100));
    }

    #[test]
    fn char_set_class() {
      assert_debug_snapshot!(parse_regex(r"[a-zA-Z_][\w_]*", 100));
    }

    #[test]
    fn char_set_escape() {
      assert_debug_snapshot!(parse_regex(r"([a-zA-\U{45}_] )", 100));
    }

    #[test]
    fn char_set_hyphen() {
      assert_debug_snapshot!(parse_regex(r"([-[-z ---_-] )", 100));
    }

    #[test]
    fn repetition_mixed() {
      assert_debug_snapshot!(parse_regex(r"\d+(\.\d*)?", 100));
    }
  }
}