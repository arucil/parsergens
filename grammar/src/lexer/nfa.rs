use std::hash::Hash;
use std::fmt::{self, Debug};
use super::dfa::Dfa;
use super::nfa_builder::NfaBuilder;
use crate::{Map, Set};

#[derive(Debug)]
pub struct Nfa<A, P, V> {
  pub transitions: Map<(State, Option<A>), Set<State>>,
  pub alphabets: Map<State, Set<A>>,
  pub accept_states: Map<State, (P, V)>,
}

#[derive(PartialEq, Eq, Hash, Clone, Copy)]
pub struct State(pub usize);

impl<A, P, V> Nfa<A, P, V>
  where A: Eq + Hash + Copy,
        P: PartialOrd
{
  pub fn builder() -> NfaBuilder<A, P, V> {
    NfaBuilder::new()
  }
}

impl<A, P, V> Nfa<A, P, V>
  where A: Eq + Hash + Copy + Debug,
        P: PartialOrd,
        V: Clone
{
  pub fn to_dfa(&self, start: State) -> Dfa<A, V> {
    super::powerset_cons::powerset(self, start)
  }
}

impl Debug for State {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    write!(f, "State({})", self.0)
  }
}