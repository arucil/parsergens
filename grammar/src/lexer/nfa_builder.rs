use std::hash::Hash;
use super::nfa::*;
use super::{Map, Set};

pub struct NfaBuilder<A, P, V> {
  counter: usize,
  transitions: Map<(State, Option<A>), Set<State>>,
  alphabets: Map<State, Set<A>>,
  accept_states: Map<State, (P, V)>,
}

impl<A, P, V> NfaBuilder<A, P, V>
  where A: Eq + Hash + Copy,
        P: PartialOrd
{
  pub fn new() -> Self {
    Self {
      counter: 0,
      transitions: Map::new(),
      alphabets: Map::new(),
      accept_states: Map::new(),
    }
  }

  pub fn build(self) -> Nfa<A, P, V> {
    Nfa {
      transitions: self.transitions,
      alphabets: self.alphabets,
      accept_states: self.accept_states,
    }
  }

  pub fn state(&mut self) -> State {
    self.counter += 1;
    State(self.counter)
  }

  pub fn transition(&mut self, src: State, dest: State, c: Option<A>) {
    self.transitions.entry((src, c)).or_default().insert(dest);
    if let Some(c) = c {
      self.alphabets.entry(src).or_default().insert(c);
    }
  }

  pub fn accept(&mut self, state: State, priority: P, value: V) {
    self.accept_states.insert(state, (priority, value));
  }
}