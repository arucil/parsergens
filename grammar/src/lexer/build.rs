use std::collections::BTreeSet;
use super::super::grammar_parser::ast::*;
use super::super::grammar_parser::regex::{RegexError, RegexErrorKind};
use super::nfa::{State, Nfa};
use super::nfa_builder::NfaBuilder;
use super::{Lexer, TokenId, LexerError};
use super::util;
use crate::{Set, BiMap};

pub fn build(decls: &[&TokenDecl], skips: &[&SkipDecl]) -> Result<Lexer, LexerError> {
  if decls.is_empty() {
    return Err(LexerError::NoTokens);
  }

  validate_decls(decls, skips)?;

  let char_intervals = compute_char_intervals(decls, skips);

  let mut nfa_builder = Nfa::builder();
  let start = nfa_builder.state();
  let mut token_gen = TokenGen::default();
  let mut token_names = BiMap::new();

  for decl in decls {
    let accept = add_regex_to_nfa(
      &mut nfa_builder, &decl.pattern.regex, start, &char_intervals);
    let token = token_gen.gen();
    token_names.insert(token, decl.name.1.clone());

    nfa_builder.accept(accept, decl.pattern.source.0 .0, token);
  }

  for decl in skips {
    let accept = add_regex_to_nfa(
      &mut nfa_builder, &decl.pattern.regex, start, &char_intervals);
    let token = token_gen.gen();

    nfa_builder.accept(accept, decl.pattern.source.0 .0, token);
  }

  let nfa = nfa_builder.build();
  let dfa = nfa.to_dfa(start).minimize();

  Ok(Lexer {
    dfa,
    char_intervals,
    token_names,
  })
}

fn add_regex_to_nfa(
  builder: &mut NfaBuilder<u32, usize, TokenId>,
  regex: &Regex,
  enter: State,
  char_intervals: &[u32],
) -> State {
  match regex {
    Regex::Empty => {
      enter
    }
    Regex::Any => {
      let exit = builder.state();
      add_intervals_to_nfa(builder, enter, exit, 0, char_intervals.len() as u32);
      exit
    }
    Regex::Char(c) => {
      let exit = builder.state();
      let interval = util::find_char_interval(*c as u32, char_intervals);
      builder.transition(enter, exit, Some(interval));
      exit
    }
    Regex::CharSet(set, negated) => {
      let exit = builder.state();
      if *negated {
        let mut neg_set = (0..char_intervals.len() as u32).collect::<Set<_>>();
        for item in set {
          match item {
            CharSetItem::Char(c) => {
              let interval = util::find_char_interval(*c as u32, char_intervals);
              neg_set.remove(&interval);
            }
            CharSetItem::CharClass(class) => {
              apply_char_class(*class, char_intervals, |lower, upper| {
                for i in lower..upper {
                  neg_set.remove(&i);
                }
              });
            }
            CharSetItem::Range(a, b) => {
              let lower = util::find_char_interval(*a as u32, char_intervals);
              let upper = util::find_char_interval(*b as u32 + 1, char_intervals);
              for i in lower..upper {
                neg_set.remove(&i);
              }
            }
          }
        }

        for interval_ix in neg_set {
          builder.transition(enter, exit, Some(interval_ix));
        }
      } else {
        for item in set {
          match item {
            CharSetItem::Char(c) => {
              let interval_ix = util::find_char_interval(*c as u32, char_intervals);
              builder.transition(enter, exit, Some(interval_ix));
            }
            CharSetItem::CharClass(class) => {
              add_char_class_to_nfa(builder, enter, exit, *class, char_intervals);
            }
            CharSetItem::Range(a, b) => {
              let lower = util::find_char_interval(*a as u32, char_intervals);
              let upper = util::find_char_interval(*b as u32 + 1, char_intervals);
              add_intervals_to_nfa(builder, enter, exit, lower, upper);
            }
          }
        }
      }
      exit
    }
    Regex::CharClass(class) => {
      let exit = builder.state();
      add_char_class_to_nfa(builder, enter, exit, *class, char_intervals);
      exit
    }
    Regex::Alt(alts) => {
      let exit = builder.state();
      for alt in alts {
        let last = add_regex_to_nfa(builder, alt, enter, char_intervals);
        builder.transition(last, exit, None);
      }
      exit
    }
    Regex::Concat(items) => {
      items.iter().fold(enter, |enter, item| {
        add_regex_to_nfa(builder, item, enter, char_intervals)
      })
    }
    Regex::Optional(item) => {
      let exit = add_regex_to_nfa(builder, item, enter, char_intervals);
      builder.transition(enter, exit, None);
      exit
    }
    Regex::Many(item) => {
      let exit = builder.state();
      builder.transition(enter, exit, None);
      let temp = add_regex_to_nfa(builder, item, exit, char_intervals);
      builder.transition(temp, exit, None);
      exit
    }
    Regex::Many1(item) => {
      let temp = builder.state();
      builder.transition(enter, temp, None);
      let exit = add_regex_to_nfa(builder, item, temp, char_intervals);
      builder.transition(exit, temp, None);
      exit
    }
  }
}

fn add_char_class_to_nfa(
  builder: &mut NfaBuilder<u32, usize, TokenId>,
  enter: State,
  exit: State,
  class: CharClass,
  char_intervals: &[u32]
) {
  apply_char_class(class, char_intervals, |lower, upper| {
    add_intervals_to_nfa(builder, enter, exit, lower, upper);
  });
}

fn apply_char_class<F>(
  class: CharClass,
  char_intervals: &[u32],
  mut action: F
)
  where F: FnMut(u32, u32)
{
  for (c0, c1) in class.ranges() {
    let lower = util::find_char_interval(c0 as u32, char_intervals);
    let upper = util::find_char_interval(c1 as u32 + 1, char_intervals);
    action(lower, upper);
  }
}

fn add_intervals_to_nfa(
  builder: &mut NfaBuilder<u32, usize, TokenId>,
  enter: State,
  exit: State,
  lower_ix: u32,
  upper_ix: u32
) {
  for i in lower_ix..upper_ix {
    builder.transition(enter, exit, Some(i));
  }
}

fn compute_char_intervals(
  decls: &[&TokenDecl],
  skips: &[&SkipDecl]
) -> Vec<u32> {
  let mut char_intervals = BTreeSet::new();
  char_intervals.insert(0);

  for decl in decls {
    collect_char_intervals(&decl.pattern.regex, &mut char_intervals);
  }

  for decl in skips {
    collect_char_intervals(&decl.pattern.regex, &mut char_intervals);
  }

  char_intervals.into_iter().collect()
}

fn collect_char_intervals(regex: &Regex, intervals: &mut BTreeSet<u32>) {
  match regex {
    Regex::Any => {}
    Regex::Empty => {}
    Regex::Char(c) => {
      intervals.insert(*c as u32);
      intervals.insert(*c as u32 + 1);
    }
    Regex::CharSet(set, _) => {
      for item in set {
        match item {
          CharSetItem::Char(c) => {
            intervals.insert(*c as u32);
            intervals.insert(*c as u32 + 1);
          }
          CharSetItem::CharClass(class) => {
            collect_char_class_intervals(class, intervals);
          }
          CharSetItem::Range(a, b) => {
            intervals.insert(*a as u32);
            intervals.insert(*b as u32 + 1);
          }
        }
      }
    }
    Regex::CharClass(class) => {
      collect_char_class_intervals(class, intervals);
    }
    Regex::Alt(items) | Regex::Concat(items) => {
      for item in items {
        collect_char_intervals(item, intervals);
      }
    }
    Regex::Optional(item) | Regex::Many(item) | Regex::Many1(item) => {
      collect_char_intervals(item, intervals);
    }
  }
}

fn collect_char_class_intervals(class: &CharClass, intervals: &mut BTreeSet<u32>) {
  for (c0, c1) in class.ranges() {
    intervals.insert(c0 as u32);
    intervals.insert(c1 as u32 + 1);
  }
}

fn validate_decls(
  decls: &[&TokenDecl],
  skips: &[&SkipDecl]
) -> Result<(), LexerError> {
  for decl in decls {
    if regex_accepts_empty_string(&decl.pattern.regex) {
      return Err(LexerError::RegexError(RegexError {
        kind: RegexErrorKind::Empty,
        span: decl.pattern.source.0,
      }));
    }
  }

  for decl in skips {
    if regex_accepts_empty_string(&decl.pattern.regex) {
      return Err(LexerError::RegexError(RegexError {
        kind: RegexErrorKind::Empty,
        span: decl.pattern.source.0,
      }));
    }
  }

  Ok(())
}

fn regex_accepts_empty_string(regex: &Regex) -> bool {
  match regex {
    Regex::Empty => true,
    Regex::Alt(alts) => alts.iter().any(regex_accepts_empty_string),
    Regex::Concat(items) => items.iter().all(regex_accepts_empty_string),
    Regex::Many(_) | Regex::Optional(_) => true,
    Regex::Many1(item) => regex_accepts_empty_string(item),
    Regex::Any | Regex::Char(_) | Regex::CharClass(_) => false,
    Regex::CharSet(set, negated) => {
      if *negated {
        // TODO this needs further consideration
        false
      } else {
        char_set_accepts_empty_string(set)
      }
    }
  }
}

fn char_set_accepts_empty_string(set: &[CharSetItem]) -> bool {
  set.iter()
    .all(|item| {
      match item {
        CharSetItem::Char(_) => false,
        CharSetItem::CharClass(_) => false,
        CharSetItem::Range(a, b) => a > b,
      }
    })
}

#[derive(Default)]
struct TokenGen(u32);

impl TokenGen {
  fn gen(&mut self) -> TokenId {
    let i = self.0;
    self.0 += 1;
    TokenId(i)
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use super::super::super::grammar_parser;
  use insta::assert_debug_snapshot;

  #[test]
  fn longest_match() {
    let ast = grammar_parser::parse(r#"
%token IF "if"
%token IN "int"
%token ID /[a-zA-Z_][\w_]*/
    "#).unwrap();

    let decls = ast.iter()
      .filter_map(|decl| match &decl.1 {
        Decl::Token(decl) => Some(decl),
        _ => None,
      })
      .collect::<Vec<_>>();

    let char_intervals = compute_char_intervals(&decls, &[]);

    let mut nfa_builder = NfaBuilder::new();
    let start = nfa_builder.state();
    let mut token_gen = TokenGen::default();

    for decl in decls {
      let accept = add_regex_to_nfa(
        &mut nfa_builder, &decl.pattern.regex, start, &char_intervals);
      let token = token_gen.gen();

      nfa_builder.accept(accept, !decl.pattern.source.0 .0, token);
    }

    let nfa = nfa_builder.build();
    let dfa = nfa.to_dfa(start);

    assert_debug_snapshot!((
      char_intervals.iter().enumerate().collect::<Vec<_>>(),
      nfa,
      dfa,
    ));
  }

  #[test]
  fn regex_repetition() {
    let ast = grammar_parser::parse(r#"
%token INT /\d+(\.\d*)?/
    "#).unwrap();

    let decls = ast.iter()
      .filter_map(|decl| match &decl.1 {
        Decl::Token(decl) => Some(decl),
        _ => None,
      })
      .collect::<Vec<_>>();

    let char_intervals = compute_char_intervals(&decls, &[]);

    let mut nfa_builder = NfaBuilder::new();
    let start = nfa_builder.state();
    let mut token_gen = TokenGen::default();

    for decl in decls {
      let accept = add_regex_to_nfa(
        &mut nfa_builder, &decl.pattern.regex, start, &char_intervals);
      let token = token_gen.gen();

      nfa_builder.accept(accept, !decl.pattern.source.0 .0, token);
    }

    let nfa = nfa_builder.build();
    let dfa = nfa.to_dfa(start);

    assert_debug_snapshot!((
      char_intervals.iter().enumerate().collect::<Vec<_>>(),
      nfa,
      dfa,
    ));
  }

  #[test]
  fn line_comment() {
    let ast = grammar_parser::parse(r#"
%token COMMENT /#[^\n]*/
    "#).unwrap();

    let decls = ast.iter()
      .filter_map(|decl| match &decl.1 {
        Decl::Token(decl) => Some(decl),
        _ => None,
      })
      .collect::<Vec<_>>();

    let char_intervals = compute_char_intervals(&decls, &[]);

    let mut nfa_builder = NfaBuilder::new();
    let start = nfa_builder.state();
    let mut token_gen = TokenGen::default();

    for decl in decls {
      let accept = add_regex_to_nfa(
        &mut nfa_builder, &decl.pattern.regex, start, &char_intervals);
      let token = token_gen.gen();

      nfa_builder.accept(accept, !decl.pattern.source.0 .0, token);
    }

    let nfa = nfa_builder.build();
    let dfa = nfa.to_dfa(start);

    assert_debug_snapshot!((
      char_intervals.iter().enumerate().collect::<Vec<_>>(),
      nfa,
      dfa,
    ));
  }
}