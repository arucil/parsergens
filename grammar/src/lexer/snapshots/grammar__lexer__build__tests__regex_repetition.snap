---
source: grammar/src/lexer/build.rs
expression: "(char_intervals.iter().enumerate().collect::<Vec<_>>(), nfa, dfa)"
---
(
    [
        (
            0,
            0,
        ),
        (
            1,
            46,
        ),
        (
            2,
            47,
        ),
        (
            3,
            48,
        ),
        (
            4,
            58,
        ),
    ],
    Nfa {
        transitions: {
            (
                State(0),
                None,
            ): {
                State(1),
            },
            (
                State(1),
                Some(
                    3,
                ),
            ): {
                State(2),
            },
            (
                State(2),
                None,
            ): {
                State(1),
                State(4),
            },
            (
                State(2),
                Some(
                    1,
                ),
            ): {
                State(3),
            },
            (
                State(3),
                None,
            ): {
                State(4),
            },
            (
                State(4),
                Some(
                    3,
                ),
            ): {
                State(5),
            },
            (
                State(5),
                None,
            ): {
                State(4),
            },
        },
        alphabets: {
            State(1): {
                3,
            },
            State(2): {
                1,
            },
            State(4): {
                3,
            },
        },
        accept_states: {
            State(4): (
                18446744073709551603,
                TokenId(
                    0,
                ),
            ),
        },
    },
    Dfa {
        start: 0,
        transitions: {
            (
                State(0),
                3,
            ): State(1),
            (
                State(1),
                3,
            ): State(2),
            (
                State(1),
                1,
            ): State(3),
            (
                State(2),
                3,
            ): State(2),
            (
                State(2),
                1,
            ): State(3),
            (
                State(3),
                3,
            ): State(4),
            (
                State(4),
                3,
            ): State(4),
        },
        accept_states: {
            State(1): TokenId(
                0,
            ),
            State(2): TokenId(
                0,
            ),
            State(3): TokenId(
                0,
            ),
            State(4): TokenId(
                0,
            ),
        },
    },
)
